lib/spack/spack/variant.py

```diff
@@ -617,6 +623,102 @@ def substitute_abstract_variants(spec):
         spec.variants.substitute(new_variant)
 
 
+class DisjointSetsOfValues(Sequence):
+    """Allows combinations from one of among many mutually exclusive sets.
+
+    Args:
+        *sets (list of tuples): mutually exclusive sets of values
+    """
+    def __init__(self, *sets):
+        self.sets = [set(x) for x in sets]
+        if any(s1 & s2 for s1, s2 in itertools.combinations(self.sets, 2)):
+            raise error.SpecError('sets in input must be disjoint')
+
+        self.feature_values = tuple(itertools.chain.from_iterable(self.sets))
+        self.default = None
+        self.multi = True
+        self.error_fmt = "variant '{0}' in package '{1}' accepts " \
+                         "combinations from a single disjoint set of " \
+                         "values [{2}]"
+
+    def with_default(self, default):
+        """Sets the default value and returns self."""
+        self.default = default
+        return self
+
+    def with_error(self, error_fmt):
+        """Sets the error message format and returns self."""
+        self.error_fmt = error_fmt
+        return self
+
+    def with_non_feature_values(self, *values):
+        """Marks a few values as not being tied to a feature."""
+        self.feature_values = tuple(
+            x for x in self.feature_values if x not in values
+        )
+        return self
+
+    def __contains__(self, item):
+        return item in itertools.chain.from_iterable(self.sets)
+
+    def __getitem__(self, idx):
+        return tuple(self.feature_values)[idx]
+
+    def __len__(self):
+        return len(self.feature_values)
+
+    @property
+    def validator(self):
+        def _disjoint_set_validator(pkg_name, variant_name, values):
+            # If for any of the sets, all the values are in it return True
+            if any(all(x in s for x in values) for s in self.sets):
+                return
+            msg = self.error_fmt.format(variant_name, pkg_name, values)
+            raise error.SpecError(msg)
+        return _disjoint_set_validator
+
+
+def _a_single_value_or_a_combination(single_value, *values):
+    error = "the value '" + single_value + \
+            "' in the variant '{0}' of package '{1}' " \
+            "is mutually exclusive with any of the other values"
+    return DisjointSetsOfValues(
+        (single_value,), values
+    ).with_default(single_value).with_error(error).\
+        with_non_feature_values(single_value)
+
+
+def any_combination_of(*values):
```

Comment:
To me `any_combination_of` includes the empty set, which is represented by `none`. Can we just improve the docstring of this function in case? I'd like to keep the names as brief as possible (without sacrificing clarity).

Reply:
Perhaps this function (or both functions) could have an `_allow_empty_set` option which defaults to `True` and adds the `"none"` option.

`auto_or_any_combination_of` does not include "none", so I think including that parameter would be a good compromise between keeping clear but concise function names and naming consistency.